### DISCUSSION TITLE
Streaming Writer

### DISCUSSION THREAD

**Discussion by sandstrom** [ohler55/ox#273] [Ideas]
Just wanted to raise the idea of a streaming writer, where data is sequentially committed to a file (or other Ruby IO object).

Basically, instead of building a complete in-memory model of the XML file, before calling `Ox.dump` on it, one would write directly to the underlying IO object (Ruby IO has built-in buffering).

Ox would still need to keep in memory the number of open tags. For some types of documents this would be a lot of data (very deep XML trees). But for most XML documents, you are rarely more than 5-10 layers deep at any one point in time, so in these cases the amount of data held in memory should be minimal.

Just wanted to put this idea out there.

I'm not saying this is something that should be added right now, but if others also find it useful, it could be explored further.

The primary use-case would be large XML exports, e.g. in the hundreds of mb, where you'd prefer to not hold everything in memory.

**Comment by ohler55** [comment]
For such a feature I'd suggest the Ox.Builder could either be extended to support streams or a new class with the same methods could be used. What do you think of the Builder API?

**Reply by sandstrom** [reply]
I haven't actually seen the builder API, is it documented anywhere? (I'm probably blind)

**Reply by ohler55** [reply]
That's odd. For some reason the docs are being generated for that class. I'll have to figure that out. If you look at the ext/ox/builder.c code you can see a bit more about it.

**Reply by ohler55** [reply]
Fixed the doc problem. Please take a look at the Ox::Builder API as a model for a streaming builder.

**Reply by sandstrom** [reply]
I ended up rolling our own streaming XML writer. The Ox::Builder looks quite similar to what we ended up with.

The only difference is that our `element` method takes a block (calling `element` again, inside the block, will render a child-node), so there is no manual closing of elements. But other than that, it looks very similar!

I should mention that the manual closing of elements could actually be very good in the case of a library (which Ox is), since that makes it easier to adopt Ox in your application (e.g. if you need a custom layer on top), so I'm not saying there is anything bad with that.

Ours isn't in C though, so I'm sure Ox::Builder would be quicker.

**Reply by ohler55** [reply]
It's been so long since I looked at the builder I didn't remember that the Ox::Builder already support writing to a file or to any io object. Does that work for you?

**Reply by sandstrom** [reply]
Yeah, I think that'll probably do the job! 😄 

Haven't looked at it since we ended up writing it ourselves, but most likely.

**Reply by dvandersluis** [reply]
@ohler55 where can I find the documentation for `Ox::Builder`?

**Reply by ohler55** [reply]
http://ohler.com/ox/Ox/Builder.html

**Reply by dvandersluis** [reply]
Thanks! :raised_hands: How do I use it to stream to a file? (Is there more to it than `Ox::Builder.file`?)

**Reply by ohler55** [reply]
http://ohler.com/ox/Ox/Builder.html#method-c-file
